File: sorting/gnome.c
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "array.h"

void gnome_sort(Arr_T arr) {
  int pos = 0;

  while (pos < arr.size) {
    if ((pos == 0) || (arr.arr[pos] >= arr.arr[pos - 1])) {
      pos += 1;
    } else {
      swap(pos, pos - 1, arr);
      pos -= 1;
    }
  }

  return;
}

void sort(Arr_T arr) { gnome_sort(arr); }

#ifndef ASLIB
int main(int argc, char *argv[]) {
  int arr_size;
  if (argc > 1) {
    if (sscanf(argv[1], "%i", &arr_size) != 1) {
      fprintf(stderr, "[ERR] - Not an integer!");
      return 1;
    }
  } else {
    arr_size = 100;
  }
  Arr_T M = make_Arr(arr_size);
  populate_Arr(M);
  printf("Array before sorting: \n");
  print_Arr(M);
  combSort(M);
  printf("Array after sorting: \n");
  print_Arr(M);
  return 0;
}
#endif

File: sorting/gnome.py
import array


def gnome_sort(lst):
    pos = 0

    while pos < lst.size:
        if (pos == 0) or (lst.arr[pos] >= lst.arr[pos - 1]):
            pos += 1
        else:
            lst.swap(pos, pos-1)
            pos -= 1


if __name__ == "__main__":
    array = array.Arr_T()
    array.make_Arr(100)
    array.populate_Arr()
    print("Array before sorting: ")
    array.print_Arr()
    gnome_sort(array)
    print("Array after sorting: ")
    array.print_Arr()